### PR TITLE
Build OpenRCT2 with breakpad support in CI

### LINKED ISSUE #17632: Crash dumps are no longer made like before and not uploaded to backtrace
### Operating System

Windows 10, 64 bit

### OpenRCT2 build

OpenRCT2, v0.4.1 Release (be518f4) provided by GitHub

### Describe the issue

When the game crashes, it used to create a crash dump, save file, screenshot, copy of the config, and ask the player to automatically upload them to Backtrace for inspection. This no longer happens. Instead, crash dumps now appear in `%localappdata%\CrashDumps`, and without the other files.

### Area(s) with issue?

_No response_

### Steps to reproduce

1. Start the game
2. Force a game crash by running the `dereference` game command
3. Observe how the game just quits with no message, and check `%localappdata%\CrashDumps`

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by janisozaur**
Fixes #17632

### CODE REVIEW

**Comment by janisozaur** [comment]
I have not actually tested it, just set the param. Can some Windows folk verify this?

**Comment by Broxzier** [comment]
Running the `dereference` command still results in the crash dumps being generated in local appdata and no message pops up for the user.

**Comment by janisozaur** [comment]
I tested build from CI for this commit and it worked as expected:

image

https://github.com/OpenRCT2/OpenRCT2/issues/17640

**Comment by Broxzier** [comment]
The build I used also came from CI:

https://user-images.githubusercontent.com/9705046/181827289-0991ab48-8791-4f79-ae9d-b4f4a8f90ce5.mp4

Why wouldn't it show me the message with the same build?

**Comment by Broxzier** [comment]
There are files being created in the `Documents\OpenRCT2\crash` folder, but no pop-is shown afterwards for uploading the files. No files are created by the current release build, so that is at least shows some progress.

**Comment by janisozaur** [comment]
anything interesting in the logs?

**Comment by Broxzier** [comment]
Only this in stdout:

```
openrct2 $ Dump Path: C:\Users\User\Documents\OpenRCT2\crash
Dump File Path: C:\Users\User\Documents\OpenRCT2\crash\599f4304-82fa-4e8b-be16-ad1bc3223859(6c7c006_x86-64).dmp
Dump Id: 599f4304-82fa-4e8b-be16-ad1bc3223859
Version: 0.4.1
Commit: 6c7c006
```

**Comment by janisozaur** [comment]
Ugh, I haven't noticed this PR was made from origin, not my fork.

**Comment by Broxzier** [comment]
Hahah, well that explains the difference in behaviour. I will give this another spin once CI is green.

**Comment by janisozaur** [comment]
A recent (today's) backtrace bug report: https://github.com/OpenRCT2/OpenRCT2/issues/17651

**Comment by Broxzier** [comment]
That was is from an older release, OpenRCT2, v0.4.0 (c6302a1 on master) provided by GitHub.

File: scripts/build [hunk 1]
```diff
@@ -26,17 +26,21 @@ if [[ "$OSTYPE" == "cygwin" || "$OSTYPE" == "msys" ]]; then
...
     vstool editbin //subsystem:windows bin/openrct2.exe
+
+    bin/openrct2.com --version
 else
     echo -e "\033[0;36mBuilding OpenRCT2...\033[0m"
     mkdir -p bin && cd bin
     export DESTDIR=install
     cmake .. -G Ninja -DCMAKE_INSTALL_PREFIX=/usr "$@"
 
+    ./openrct2 --version
```
Comment (Broxzier) [inline scripts/build:43]: This needs to move down to after compilation

**Comment by Broxzier** [comment]
I'm still getting no message asking me to upload the crash dumps. The dumps are created in both the `crash` folder and the `%localappdata%\CrashDump` folder.

```
> .\openrct2.com --version
OpenRCT2, v0.4.1-81-gf48d485 (f48d485 on janisozaur-patch-1) provided by GitHub
Windows (x86-64)
Network version: 0.4.1-2
Plugin API version: 57
Current park file version: 13
Minimum park file version: 12
With breakpad support enabled
```

On crash I see:

```
Silent replay recording started: (C:\Users\User\Documents\OpenRCT2\debug_replay.parkrep) C:\Users\User\Documents\OpenRCT2\debug_replay.parkrep

Replay recording stopped: (C:\Users\User\Documents\OpenRCT2\debug_replay.parkrep) C:\Users\User\Documents\OpenRCT2\debug_replay.parkrep
  Ticks: 218
  Commands: 0
  Checksums: 6
openrct2 $ Dump Path: C:\Users\User\Documents\OpenRCT2\crash
Dump File Path: C:\Users\User\Documents\OpenRCT2\crash\59e0f31b-8701-4caa-9e40-753c9d575a2c(f48d485_x86-64).dmp
Dump Id: 59e0f31b-8701-4caa-9e40-753c9d575a2c
Version: 0.4.1
Commit: f48d485
```

**Comment by janisozaur** [comment]
Oh well, I've resorted to adding quite a few `printf`s, please provide logs @Broxzier

**Comment by Broxzier** [comment]
I had to cancel the macOS jobs, because after over 2 hours they were still holding up the artifacts...

Anyway, here is everything that gets printed:

```
PS C:\Users\User\Documents\OpenRCT2\bin-testing> .\openrct2.com --version
OpenRCT2, v0.4.1-86-gf527e5b (f527e5b on refs/pull/17634/merge) provided by GitHub
Windows (x86-64)
Network version: 0.4.1-2
Plugin API version: 57
Current park file version: 13
Minimum park file version: 12
With breakpad support enabled
PS C:\Users\User\Documents\OpenRCT2\bin-testing> .\openrct2.com
janis: D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:358
janis: D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:340
janis: D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:342
janis: D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:344
janis: D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:362
openrct2 $ janis: starting OnCrash D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:103
janis: D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:139
janis: compressing dump D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:143
janis: dump compressed D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:149
janis: rename files D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:168
Dump Path: C:\Users\User\Documents\OpenRCT2\crash
Dump File Path: C:\Users\User\Documents\OpenRCT2\crash\457091d3-8a3a-4edb-b7a7-e07afe3e2ac7(f527e5b_x86-64).dmp
Dump Id: 457091d3-8a3a-4edb-b7a7-e07afe3e2ac7
Version: 0.4.1
Commit: f527e5b
janis: start dumping map D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:184
janis: dump map start D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:187
janis: map prepared D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:189
janis: exporter prepared D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:193
janis: context obrtained D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:195
janis: obj manager obtained D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:197
janis: packable object list done D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:199
janis: save exported D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:202
janis: map dump completed D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:209
janis: D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:214
janis: saving config D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:219
janis: config saved D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:222
janis: saving screenshot D:\a\OpenRCT2\OpenRCT2\src\openrct2\platform\Crash.cpp:226
```

**Comment by janisozaur** [comment]
Ok, thanks. I kind of suspected screenshot might be the problem, as I recall we received some bug reports about that recently. Still surprised about it. What display engine are you using and does changing it produce any different results?

**Comment by Broxzier** [comment]
That indeed looks to be the culprit. I was on OpenGL. Changing this to software made it work properly.

**Comment by janisozaur** [comment]
Perhaps the gl context is gone by the time we make it current for rendering. Will have to look into that part of code a bit. Is this pr necessary or does breakpad work correctly even without the change to msbuild in CI?

**Comment by Broxzier** [comment]
Yes, this PR is necessary. In the current release, no crashdumps are created and uploaded at all.

**Comment by IntelOrca** [comment]
Probably a regression from https://github.com/OpenRCT2/OpenRCT2/pull/16887

@janisozaur This is how msbuild properties work:

* All environment variables are available as properties in the .proj.
* These are overriden by anything specificed on the command line.
* Finally they may be set or overriden in the actual .proj or an import.

I wasn't sure whether `msbuild` accepted `-p`, you have a mix or using `/` and `-` for the switches.

**Comment by janisozaur** [comment]
https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference?view=vs-2022 lists all options as using `-`. I agree on the potential regression cause - I've mentioned it here: https://github.com/OpenRCT2/OpenRCT2/issues/17632#issuecomment-1198637893

Right now I'm having CI build various test versions so I can figure out what best to do about breakpad not presenting the usual dialogue when using OpenGL. The latest iteration significantly ups the required version and doesn't provide any more info on context getting possibly lost (at least when I test in on wine). I think I will stop now and use one of the following approaches:
1. generate screenshot with software rather than opengl - this might require allocations and not provide accurate information
2. not generate screenshot at all.

**Comment by IntelOrca** [comment]
> https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference?view=vs-2022 lists all options as using -. I agree on the potential regression cause - I've mentioned it here: https://github.com/OpenRCT2/OpenRCT2/issues/17632#issuecomment-1198637893

I am suggesting you be consistent, change `/t` to `-t`.

**Comment by janisozaur** [comment]
One more test I've done for this was trying to implement a screenshot using Windows' native APIs - my reasoning was that perhaps the OS still holds on to data that it needs to display even when covered with 'crashed' overlay, but I failed to obtain meaningful results/

The PoC patch is here:
```diff
diff --git a/src/openrct2-ui/CMakeLists.txt b/src/openrct2-ui/CMakeLists.txt
index 4cafef445..65839095f 100644
--- a/src/openrct2-ui/CMakeLists.txt
+++ b/src/openrct2-ui/CMakeLists.txt
@@ -111,7 +111,7 @@ if (WIN32)
     # mingw complains about "%zu" not being a valid format specifier for printf, unless we
     # tell it that it is
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -D__USE_MINGW_ANSI_STDIO=1")
-    target_link_libraries(${PROJECT_NAME} comdlg32)
+    target_link_libraries(${PROJECT_NAME} comdlg32 gdiplus)
     if (MSVC)
         target_link_libraries(${PROJECT_NAME} SDL2::SDL2-static)
         target_include_directories(${PROJECT_NAME} SYSTEM PRIVATE SDL2::SDL2-static)
diff --git a/src/openrct2-ui/UiContext.Win32.cpp b/src/openrct2-ui/UiContext.Win32.cpp
index c8915c19a..546e4b55a 100644
--- a/src/openrct2-ui/UiContext.Win32.cpp
+++ b/src/openrct2-ui/UiContext.Win32.cpp
@@ -14,6 +14,7 @@
 #    include <windows.h>
 #    include <shellapi.h>
 #    include <commdlg.h>
+#include <gdiplus.h>
 // clang-format on
 #    undef CreateWindow
 
@@ -211,6 +212,69 @@ namespace OpenRCT2::Ui
             return true;
         }
 
+        void native_screenshot(SDL_Window* window) override
+        {
+            printf("janis doing screenshot!\n");
+            auto hwnd = GetHWND(window);
+            auto hdcSrc = GetWindowDC(hwnd);
+            RECT window_rect{};
+            GetWindowRect(hwnd, &window_rect);
+            auto hdcDest = CreateCompatibleDC(hdcSrc);
+            auto width = window_rect.right - window_rect.left;
+            auto height = window_rect.bottom - window_rect.top;
+            // create a bitmap we can copy it to,
+            // using GetDeviceCaps to get the width/height
+            auto hBitmap = CreateCompatibleBitmap(hdcSrc, width, height);
+            // select the bitmap object
+            auto hOld = SelectObject(hdcDest, hBitmap);
+            // bitblt over
+            BitBlt(hdcDest, 0, 0, width, height, hdcSrc, 0, 0, SRCCOPY);
+            // restore selection
+            SelectObject(hdcDest, hOld);
+            // clean up
+            DeleteDC(hdcDest);
+            ReleaseDC(hwnd, hdcSrc);
+
+            Gdiplus::Bitmap bmp(hBitmap, static_cast<HPALETTE>(0));
+            CLSID pngClsid;
+            auto GetEncoderClsid = [](const WCHAR* format, CLSID* pClsid) -> unsigned int {
+                using namespace Gdiplus;
+                UINT num = 0;  // number of image encoders
+                UINT size = 0; // size of the image encoder array in bytes
+
+                ImageCodecInfo* pImageCodecInfo = NULL;
+
+                GetImageEncodersSize(&num, &size);
+                if (size == 0)
+                    return -1; // Failure
+
+                pImageCodecInfo = static_cast<ImageCodecInfo*>(malloc(size));
+                if (pImageCodecInfo == NULL)
+                    return -1; // Failure
+
+                GetImageEncoders(num, size, pImageCodecInfo);
+
+                for (UINT j = 0; j < num; ++j)
+                {
+                    if (wcscmp(pImageCodecInfo[j].MimeType, format) == 0)
+                    {
+                        *pClsid = pImageCodecInfo[j].Clsid;
+                        free(pImageCodecInfo);
+                        return j; // Success
+                    }
+                }
+
+                free(pImageCodecInfo);
+                return -1; // Failure
+            };
+            auto res = GetEncoderClsid(L"image/png", &pngClsid);
+            printf("janis res = %u\n", res);
+            bmp.Save(L"C:/image.png", &pngClsid, NULL);
+
+            // free up the Bitmap object
+            DeleteObject(hBitmap);
+        }
+
     private:
         HWND GetHWND(SDL_Window* window)
         {
diff --git a/src/openrct2-ui/UiContext.cpp b/src/openrct2-ui/UiContext.cpp
index 4d02484f1..244a8986d 100644
--- a/src/openrct2-ui/UiContext.cpp
+++ b/src/openrct2-ui/UiContext.cpp
@@ -640,6 +640,11 @@ public:
         _platformUiContext->ShowMessageBox(_window, message);
     }
 
+    void native_screenshot() override
+    {
+        _platformUiContext->native_screenshot(_window);
+    }
+
     bool HasMenuSupport() override
     {
         return _platformUiContext->HasMenuSupport();
diff --git a/src/openrct2-ui/UiContext.h b/src/openrct2-ui/UiContext.h
index e057ec4a7..2ce3868ef 100644
--- a/src/openrct2-ui/UiContext.h
+++ b/src/openrct2-ui/UiContext.h
@@ -34,6 +34,7 @@ namespace OpenRCT2
             virtual bool IsSteamOverlayAttached() abstract;
 
             virtual void ShowMessageBox(SDL_Window* window, const std::string& message) abstract;
+            virtual void native_screenshot(SDL_Window* window) abstract;
             virtual bool HasMenuSupport() abstract;
             virtual int32_t ShowMenuDialog(
                 const std::vector<std::string>& options, const std::string& title, const std::string& text) abstract;
diff --git a/src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp b/src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp
index d387d81d9..dbe418d38 100644
--- a/src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp
+++ b/src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp
@@ -337,6 +337,7 @@ public:
 
     std::string Screenshot() override
     {
+        OpenRCT2::GetContext()->GetUiContext()->native_screenshot();
         const OpenGLFramebuffer& framebuffer = _drawingContext->GetFinalFramebuffer();
         framebuffer.Bind();
         framebuffer.GetPixels(_bitsDPI);
diff --git a/src/openrct2/drawing/NewDrawing.cpp b/src/openrct2/drawing/NewDrawing.cpp
index e1f3234d7..2c2472cfd 100644
--- a/src/openrct2/drawing/NewDrawing.cpp
+++ b/src/openrct2/drawing/NewDrawing.cpp
@@ -291,5 +291,6 @@ std::string screenshot_dump()
     {
         return drawingEngine->Screenshot();
     }
+    OpenRCT2::GetContext()->GetUiContext()->native_screenshot();
     return "";
 }
diff --git a/src/openrct2/drawing/X8DrawingEngine.cpp b/src/openrct2/drawing/X8DrawingEngine.cpp
index c2ba2a0aa..005ddd12d 100644
--- a/src/openrct2/drawing/X8DrawingEngine.cpp
+++ b/src/openrct2/drawing/X8DrawingEngine.cpp
@@ -259,6 +259,7 @@ void X8DrawingEngine::CopyRect(int32_t x, int32_t y, int32_t width, int32_t heig
 
 std::string X8DrawingEngine::Screenshot()
 {
+    OpenRCT2::GetContext()->GetUiContext()->native_screenshot();
     return screenshot_dump_png(&_bitsDPI);
 }
 
diff --git a/src/openrct2/ui/DummyUiContext.cpp b/src/openrct2/ui/DummyUiContext.cpp
index aa573ae30..9650633d2 100644
--- a/src/openrct2/ui/DummyUiContext.cpp
+++ b/src/openrct2/ui/DummyUiContext.cpp
@@ -94,6 +94,9 @@ namespace OpenRCT2::Ui
         {
             return false;
         }
+        void native_screenshot() override
+        {
+        }
 
         int32_t ShowMenuDialog(
             const std::vector<std::string>& options, const std::string& title, const std::string& text) override
diff --git a/src/openrct2/ui/UiContext.h b/src/openrct2/ui/UiContext.h
index 3660fc42f..af6dcaa2b 100644
--- a/src/openrct2/ui/UiContext.h
+++ b/src/openrct2/ui/UiContext.h
@@ -120,6 +120,7 @@ namespace OpenRCT2
             virtual void TriggerResize() abstract;
 
             virtual void ShowMessageBox(const std::string& message) abstract;
+            virtual void native_screenshot() abstract;
 
             virtual bool HasMenuSupport() abstract;
             // Creates a menu with a series of options, returns the index of the selected option

```

**Comment by janisozaur** [comment]
Awaiting review - can this get merged now?

File: src/openrct2/platform/Crash.cpp [hunk 2]
```diff
@@ -186,8 +187,9 @@ static bool OnCrash(
         exporter->Export(saveFilePathUTF8.c_str());
         savedGameDumped = true;
     }
-    catch (const std::exception&)
+    catch (const std::exception& e)
     {
+        printf("Failed to export save. Error: %s\n", e.what());
```
Comment (IntelOrca) [inline src/openrct2/platform/Crash.cpp:192]: Should this be printed to `stderr`?

File: src/openrct2/cmdline/RootCommands.cpp [hunk 3]
```diff
@@ -459,6 +459,10 @@ static void PrintVersion()
     Console::WriteLine();
     Console::WriteFormat("Minimum park file version: %d", OpenRCT2::PARK_FILE_MIN_VERSION);
     Console::WriteLine();
+#ifdef USE_BREAKPAD
+    Console::WriteFormat("With breakpad support enabled");
+    Console::WriteLine();
+#endif
```
Comment (IntelOrca) [inline src/openrct2/cmdline/RootCommands.cpp:465]: Perhaps also say that it isn't enabled when it isn't enabled.

File: src/openrct2/platform/Crash.cpp (hunk 2 above)
Comment (janisozaur) [reply src/openrct2/platform/Crash.cpp:192]: All other logs are printed already to stdout

File: src/openrct2/cmdline/RootCommands.cpp (hunk 3 above)
Comment (janisozaur) [reply src/openrct2/cmdline/RootCommands.cpp:468]: done

File: src/openrct2/platform/Crash.cpp (hunk 2 above)
Comment (IntelOrca) [reply src/openrct2/platform/Crash.cpp:192]: There is a whole bunch that print to stderr, is there another reason not to? Do you have a preference? Should more be printed to stderr?

Comment (janisozaur) [reply src/openrct2/platform/Crash.cpp:192]: Ok, i don't mind. In this pr or another one?

Comment (IntelOrca) [reply src/openrct2/platform/Crash.cpp:192]: I don't mind either.

**Review by IntelOrca — APPROVED** [review]

**Pull request merged**